exec-engine.md

```diff
@@ -0,0 +1,146 @@
+# L2 Execution Engine
+
+The rollup mirrors the layer abstraction that the Merge introduces to Ethereum L1:
+- Consensus: *Determine* canonical chain, propagate *blocks*, maintain *system-state*
+- Execution: *Process* canonical chain, propagate *transactions*, maintain *user-state*
+
+The consensus task is implemented by the rollup-node, deriving most functionality from L1.
+The execution task is implemented by the engine, providing an equivalent user environment as L1, with deposits.
+
+This document outlines the modifications, configuration and usage of a L1 execution engine for L2.
+
+
+## Deposit processing
+
+The Engine interfaces abstract away transaction types with [EIP-2718][eip-2718].
+
+To support rollup-functionality, processing of a new Deposit `TransactionType` is implemented by the engine,
+see [Deposit specification][deposit-spec].
+
+This type of transaction can mint L2 ETH, run EVM, 
+and introduce L1 information to enshrined contracts in the execution state.
+
+> **TODO**: implement deposit spec doc
+
+[deposit-spec]: ./deposits#transaction-type
+
+### Deposit boundaries
+
+Deposits are authenticated by the rollup-node, outside of the engine.
+
+To process deposits safely, the deposits MUST be authenticated first:
+- Ingest directly through trusted Engine API
+- Part of sync towards a trusted block-hash (trusted through previous Engine API instruction)
+
+Deposits MUST never be consumed from the transaction pool.
+*The transaction pool can be disabled in a deposits-only rollup*
+
+
+## Engine API
+
+*Note: the [Engine API][l1-api-spec] is in alpha, `v1.0.0-alpha.5`.
+There may be subtle tweaks, beta starts in a few weeks*
+
+### `engine_forkchoiceUpdatedV1`
+
+This updates which L2 blocks the engine considers to be canonical (`forkchoiceState` argument),
+and optionally initiates block production (`payloadAttributes` argument).
+
+Within the rollup, the types of forkchoice updates translate as:
+- `headBlockHash`: block-hash of the head of the canonical chain. Labeled `"unsafe"` in user JSON-RPC.
+   Nodes may apply L2 blocks out of band ahead of time, and then reorg when L1 data conflicts.
```

Comment:
> Reading this sequentially begs the quesiton: where are those L2 blocks coming from?

Not specified, yet. But when we move to a sequencer-enabled rollup then we can have p2p distribution of l2 blocks via rollup nodes. Without specifying that yet, I thought we'd at least need this kind of context to explain the difference between `headBlockHash` and `safeBlockHash.

The rollup-node always defines what's canonical, it never reads the engine. When there is no other input source than L1 (i.e. no p2p block propagation), then `headBlockHash == safeBlockHash`.

Reply:
In this case I would just do what I did with the rollup-node spec: say they're the same for now.

I **really, really** think we should avoid forward references, they make things much more confusing than they need to be. It's never clear if something is there for a reason or because "we'll need it later".

Reply:
In the API they are not the same, they are two distinct fields. Making these backwards changes for the deposit rollup is not going to improve anything.